### DISCUSSION TITLE
Help me dethrone PyTorch in the DAWNBench rankings!

### DISCUSSION THREAD

**Discussion by samuela** [jax-ml/jax#9669] [Q&A]
The current # 8 record holder on the [DAWNBench CIFAR-10 benchmark](https://dawn.cs.stanford.edu/benchmark/CIFAR10/train.html) is a PyTorch ResNet by @davidcpage running on an EC2 p3.2xlarge instance (single V100 GPU). Record holders # 1-# 7 either use multiple GPUs or run on more esoteric cloud providers, so I'm not worrying about them for now.

My goal is to at least match the performance of this PyTorch implementation in JAX in terms of `seconds/epoch`. However, I've found that even my somewhat carefully written JAX version is a shocking _order of magnitude slower_ than the PyTorch version. On a p3.2xlarge instance, my JAX version (attached) clocks in at about 24.2s/epoch. The PyTorch version reports completing 24 epochs in 72s, which comes out to 3s/epoch. 

Some notes:
* The entirety of the CIFAR-10 dataset is loaded into GPU memory for speeeedz.
* Training epochs are fully jit'd with `lax.scan`.
* Data augmentation is fully jit'd thanks to augmax.
* Like the pytorch version, I use float16 weights.

Original PyTorch implementation: https://github.com/davidcpage/cifar10-fast
JAX implementation (including my shell.nix for reproducibility): https://gist.github.com/samuela/78a3f0bbac759833a0464048aa499c98

What am I doing wrong here? What do I have to do to get competitive performance out of JAX?

cc @sharadmv

**Comment by YouJiacheng** [comment]
I notice that you misunderstand the usage of `jax.lax.scan`
You can check following code
```python
num_batches = 100
batch_size = 512
assert jax.lax.scan(lambda n_batches, _: (n_batches + 1, None), 0, jnp.split(jnp.zeros(num_batches * batch_size), num_batches))[0] == batch_size
```
Namely, your `num_batches` actually is the batch size and your `batch_size` actually is the number of batches.

**Reply by samuela** [reply]
Mmm I don't think so... check out the `batcher_in_paradise` function which is actually what's constructing the batches. I've confirmed that it is going through the correct number of batches in each training epoch.

**Reply by YouJiacheng** [reply]
You can just change a few lines of code to check the number of batches.

**Reply by YouJiacheng** [reply]
I think you use `jnp.split(perm[:num_batches*batch_size], num_batches)` to generate batch indices and scan on it. This make scan call the `step` function `batch_size` times.

**Reply by YouJiacheng** [reply]
```python
def make_batcher_in_paradise(num_examples: int, batch_size: int):
  assert num_examples >= batch_size
  num_batches = num_examples // batch_size
  return lambda arr: jnp.split(arr[:num_batches * batch_size], num_batches)
```
and you use the return value (a lambda) as the batch indices splitter (batcher).

**Reply by YouJiacheng** [reply]
@samuela

**Reply by samuela** [reply]
> I think you use `jnp.split(perm[:num_batches*batch_size], num_batches)` to generate batch indices and scan on it. This make scan call the `step` function `batch_size` times.

No it will make `lax.scan` call `step` `num_batches` times. Check out the docs [here](https://numpy.org/doc/stable/reference/generated/numpy.split.html). (I agree that it's confusing fwiw!)

**Reply by YouJiacheng** [reply]
There is no problem with `split`. It is `lax.scan` will treat a list of array as a `pytree`, and scan on the leading axis of each array, not the list.
You can just **RUN** my code.
Following code counts the number of calls and **ASSERT** it is equal to `batch_size`.
The assertion does  **NOT** fail.
```python
num_batches = 100
batch_size = 512
assert jax.lax.scan(lambda n_batches, _: (n_batches + 1, None), 0, jnp.split(jnp.zeros(num_batches * batch_size), num_batches))[0] == batch_size
```

**Reply by YouJiacheng** [reply]
@samuela 
https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scan.html#jax.lax.scan
> where we use [t] here to denote the type t with an additional leading axis. That is, if t is an array type then [t] represents the type with an additional leading axis, and if t is a pytree (container) type with array leaves then [t] represents the type with **the same pytree structure** and corresponding **leaves each with an additional leading axis**

https://jax.readthedocs.io/en/latest/pytrees.html
> In JAX, we use the term pytree to refer to a tree-like structure built out of container-like Python objects. Classes are considered container-like if they are in the pytree registry, which by default includes lists, tuples, and dicts. That is:.

**Reply by samuela** [reply]
Oh god you're right... That's so insanely confusing... talk about a foot-gun. Why in the world would scan ever do that?

Well good catch! I'll put together a fixed version and profile that.

**Reply by YouJiacheng** [reply]
I recommend that use `reshape` instead of `split`, that will fix it.
And you can remove the `list` to `Array` conversion if you use reshape. (since `p` will be `Array` instead of `list`)
```python
def make_batcher_in_paradise(num_examples: int, batch_size: int):
  assert num_examples >= batch_size
  num_batches = num_examples // batch_size
  return lambda arr: jnp.reshape(arr[:num_batches * batch_size], (num_batches, batch_size))
```
```python
def step(train_state, p):
    # See https://github.com/google/jax/issues/4564 as to why the array conversion is necessary.
    p = jnp.array(p) # useless with reshape
    images = ds_images[p, :, :, :]
    labels = ds_labels[p]
    images_transformed = vmap(train_transform)(augmax_rngs[p], images)
    (l, num_correct), g = value_and_grad(batch_eval, has_aux=True)(train_state.params,
                                                                   images_transformed, labels)
    return train_state.apply_gradients(grads=g), (l, num_correct)
```

**Reply by samuela** [reply]
I've fixed and updated [the gist](https://gist.github.com/samuela/78a3f0bbac759833a0464048aa499c98).

**Comment by jekbradbury** [comment]
Can you take a profile using `jax.profiler` and then upload the resulting trace.json.gz file?

**Reply by samuela** [reply]
I tried wrapping my code in `with jax.profiler.trace(log_dir="./logs")`, but I'm getting a linking error:
```
2022-02-23 19:26:39.730162: W external/org_tensorflow/tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcupti.so.11.4'; dlerror: libcupti.so.11.4: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /nix/store/5lzlfgrc37wqj400b4qgpaa6kc9d2wr1-libglvnd-1.4.0/lib:/nix/store/56si9r6nbwwyaqx1j81fn4fv8ywywmfc-nvidia-x11-510.47.03/lib:/nix/store/7ic9fxlhh2jb76b9fgdhrh1cvxm9yfzm-nvidia-x11-510.47.03-lib32/lib:/nix/store/hrc96ak11li2czfqj0xf2fc1clpn6gvx-libglvnd-1.4.0/lib::/nix/store/9lv0wxqkbqw2438wrhllcyf3sx644i5z-cudatoolkit-11.5.0/lib
```
I'm using `cuda11_cudnn82` jaxlib wheel with CUDA toolkit 11.5, cuDNN 8.3.0, and driver version 510.47.03 which is all in accordance with the [supported version constraints](https://github.com/google/jax/blob/687a7630ee77be86ec9f66e3ca5f493c4d7c5e1b/README.md?plain=1#L465) AFAIU. Am I missing something?

**Reply by hawkinsp** [reply]
That's (hopefully) a benign warning. Did a profile get produced?

**Reply by samuela** [reply]
Ah, yes I misunderstood. I've updated the gist with the new code (including profiling and the fix to the issue brought up by @YouJiacheng), and attaching the [profile.tar.gz](https://github.com/google/jax/files/8127613/profile.tar.gz) here. I did three epochs of burn-in and then profiled the training loop for 10 epochs. This was on a p3.2xlarge instance, same as the PyTorch implementation.

**Comment by jekbradbury** [comment]
Your profile shows a step time of about 237-238 ms, dominated by cuDNN operations:
![image](https://user-images.githubusercontent.com/11729078/155417496-4eace78c-7fcc-4bf5-9672-26bc4f6a0d57.png)

(For context, the annotations in the top row are names of device kernels that are called, mostly inside cuDNN; the annotations in the middle are the names of XLA ops; and the annotations in the bottom row are the kernels currently being dispatched on the host, many steps ahead of time.)

I would start by comparing the step time and cuDNN operations to a similar profile on the PyTorch side.

**Reply by jekbradbury** [reply]
Looks like most of the cuDNN kernels have theoretical occupancy of 12.5%, which seems not good. I'm not sure why that would be the case.

**Reply by samuela** [reply]
What does theoretical occupancy mean? Like simultaneous utilization of GPU cores? 

When I get a chance, I'll profile the pytorch implementation for comparison.

**Reply by samuela** [reply]
Ok, just managed to get a profile. It's 1 epoch after 3 epochs of burn-in. One thing to note is that profiling the code significantly slows down runtime: it's ~3.09s/epoch before profiling kicks in and about 12-13s/epoch with profiling. So the question remains... why is jax 10x slower?

I have to go FaceTime the parents rn, but I'm hoping to start digging into the results later this evening. Don't hesitate to pester me if you'd like the exact code and experimental details to reproduce this trace.

[pytorch-profile.tar.gz](https://github.com/google/jax/files/8145447/pytorch-profile.tar.gz)

**Reply by samuela** [reply]
I've been bumping into a few problems getting pytorch to profile the things I want and actually display them in a reasonable manner. I haven't been able to get flamegraphs working yet (GH issue forthcoming). However, I have gotten chrome-compatible traces.

I've published my testing/profiling setup in [my own fork here](https://github.com/samuela/cifar10-fast). For convenience, you can just download the saved traces [here](https://github.com/samuela/cifar10-fast/tree/master/profiling). The chrome traces can be inspect with `chrome://tracing`. For reasons I do not yet understand, `chrome_trace_3.json` seems to have captured interesting things but `chrome_trace_5.json` did not.

<img width="3096" alt="Screen Shot 2022-02-28 at 6 05 00 PM" src="https://user-images.githubusercontent.com/226872/156090996-fd4f6afc-223f-4791-af5b-2b3b11963d6e.png">

Does anyone know how I can assess the cuDNN kernel occupancy for comparison with JAX's 12.5%? I have not been able to find a comparable measurement thus far.

**Reply by samuela** [reply]
Update: flamegraph [issue created](https://github.com/pytorch/pytorch/issues/73556).

**Reply by samuela** [reply]
Just wanted to send a friendly ping on this. What can I do to get JAX to use the full cuDNN kernel occupancy available?

**Reply by mattjj** [reply]
Thanks @samuela . The last couple weeks were performance review week ("perf week") at Google so we were laden with administrative work and now we have a backlog. The pings are actually very helpful though.

**Comment by levskaya** [comment]
@samuela 
 - do you have any sense how much time is spent in your on-device augmentation vs the actual NN code?  In all "production" code using JAX we tend to use tf.data to set up the dataset augmentation pipeline since tf.data is great at utilizing CPUs.  Shoving augmentation computations onto device is only going to hurt performance since you're leaving a lot of compute on the table.
 - I think a few of us are worried about scan potentially introducing some inefficiencies here.  In practice, we never scan a training loop.  It might be innocent of trouble, but it's a bit odd.   Given jax's async dispatch you don't really gain much by doing this.

**Reply by samuela** [reply]
These are interesting points... Up to now I have been trying to `jit` as much as possible with the general assumption "XLA fast, python slow", but it sounds like this is not always the case? How should I reason about when to `jit` and when not to `jit`?

I'll try running without data augmentation as that's pretty straightforward. I can try doing a python `for` loop in place of `lax.scan` too, though personally I would consider it a bug if the `for` loop significantly is faster.

**Reply by samuela** [reply]
Just tried without any data augmentation and it's the same speed.

**Reply by samuela** [reply]
I updated the [gist](https://gist.github.com/samuela/78a3f0bbac759833a0464048aa499c98) to remove data augmentation and simplify a few other things.

**Comment by jekbradbury** [comment]
I don’t think that augmentation or overheads related to scan are
problematic here; the profile suggested that virtually all your time is
spent in the device step. But none of us on the JAX team are experts in
cuDNN, and it really feels like the issue is something like cuDNN (via XLA)
selecting slow kernels.

**Reply by levskaya** [reply]
But note this is using `augmax` which uses a bunch of on-device vmap'd `map_coordinates` calls underneath to do some affine region augmentations - you're confident those won't end up as slow kernels in the device step?

**Reply by jekbradbury** [reply]
The whole device step is highlighted in the screenshot above; it's mostly cuDNN.

**Reply by samuela** [reply]
Per @mattjj's suggestion, I tried just passing in zero inputs instead of real data. That shaved about 1 second off of the runtime, which I suspect is just due to not having to index into the data array in a non-contiguous way (plus maybe XLA is smart enough to constant fold out the first convolution layer?).

**Reply by samuela** [reply]
@jekbradbury I'm not versed in the internals of XLA/their community. Is there a way that I can raise this issue with the XLA team?

**Reply by levskaya** [reply]
@samuela - can you take your minimized code (no augmentation, just core network) and run it with the environment variable: `XLA_FLAGS=--xla_dump_to=/tmp/dir/somewhere` - this will dump the HLO graph created from your code.  We'd also need your cuda/cudnn version info and GPU make.

**Reply by samuela** [reply]
Ok, here are the [XLA dumps](https://drive.google.com/file/d/1JlIjBrd-sJoUMmyHf3jJSrLhJbUZ7Ssc/view?usp=sharing). Some version info:
* NVIDIA driver version 510.47.03
* CUDA version 11.5.0
* cuDNN version 8.3.0
* jax version 0.3.1
* jaxlib version 0.3.0
* GPU: NVIDIA Corporation GV100GL [Tesla V100 SXM2 16GB] (rev a1)
* Linux 5.13.0-1014-aws, Ubuntu, 20.04.4 LTS (Focal Fossa)

This is all running on an EC2 p3.2xlarge instance. You can use my [shell.nix](https://gist.github.com/samuela/78a3f0bbac759833a0464048aa499c98#file-shell-nix) file for complete reproducibility.

**Reply by levskaya** [reply]
@samuela - I filed an internal XLA:GPU bug about the slow kernels... hopefully someone will have some time to take a look at it.

**Reply by samuela** [reply]
Thanks @levskaya! Is there any way that I can follow along with that issue?

**Reply by levskaya** [reply]
Not really, no - but I promise that I'll update you here as soon as I know anything new.

**Reply by samuela** [reply]
Hey @levskaya, any update from the XLA folks on this?

**Reply by levskaya** [reply]
Sorry, I haven't heard anything from them yet on the issue...

**Reply by SandSnip3r** [reply]
@samuela sorry about the delay. I'm a member of the XLA GPU team and I'm on it! 😎 I'll let you know what I find.

**Reply by samuela** [reply]
Awesome, thanks so much @SandSnip3r ! Don't hesitate to reach out to me here or via email if there's anything I can do to be of assistance!

**Comment by SandSnip3r** [comment]
@samuela, I want to make absolutely sure that we're comparing apples-to-apples here. First thing I'm wondering about is the padding in the convolution. The Pytorch version uses `padding=1`, however, you do not specify a padding in your version, which appears to default to `"SAME"` in Linen. I was looking in the generated HLO, and the dimensions of the tensors change. I'm not entirely sure why that is; it smells like something related to padding, but that's not how i'd expect `"SAME"` to behave. The height and width seem to go from 32x32, to 31x31, to 30x30, to 29x29. If you want to see what I'm seeing in the HLO, open the file `module_0264.jit_train_epoch.113.before_optimizations.txt
`. What I'm specifically looking at is in the `jit_jvp_batch_eval_` module, on lines that start with something like `%convolution.327`. For example, lines 233, 246, 261, 274, etc.

Regardless, the discrepancy between the two paddings is a valid difference, right?

**Reply by samuela** [reply]
Huh that's a great point! I didn't realize that was happening... I was also under the impression that padding="SAME" in linen was the same as the padding=1 in pytorch. I'll test it out and get back to you. Could such a small change in sizes result in a large performance difference like this? (I honestly have no idea.)

**Reply by samuela** [reply]
I'm also not sure how this change in dimensions is possibly reconciled in the residual connections... I haven't check in detail but it seems like it ought to throw a shape mismatch error in that case?

**Reply by hawkinsp** [reply]
`SAME` padding in JAX has the same meaning it has in XLA and TensorFlow. i.e., it corresponds to padding on the left and right such that the output dimension has the same size as the input. If you want something different, specify the padding explicitly.

**Reply by samuela** [reply]
So IIUC padding=SAME in linen should be resulting in output arrays with the same shape as the input, but is not according to the HLOs?

**Reply by hawkinsp** [reply]
(edit: I misremembered: SAME pads on both sides)

I'm not sure. The key thing is: slightly different shapes might hit different fast paths particularly in CuDNN, so make sure you are computing exactly the same thing before comparing performance.

**Reply by SandSnip3r** [reply]
As far as what I read, SAME padding should result in the tensor shape remaining the same. I don't understand what this size change that I'm seeing in the HLO is coming from yet.

I think I'm on your side @samuela, SAME seems to be the same idea as padding=1 in the case of a 3x3 kernel. Though I do agree with @hawkinsp; a difference in shapes could have non-zero impact on performance.

**Reply by samuela** [reply]
Ok I just [instrumented the code](https://gist.github.com/samuela/6d704930b9cc5a88f149d692983f2c22) to print out shapes and the change in size appears to be derived from max-pooling not convolutions which makes some sense:
```
init             (1, 32, 32, 3)
after Conv_0     (1, 32, 32, 64)
after Conv_1     (1, 32, 32, 128)
after max_pool_0 (1, 31, 31, 128)
after Conv_2     (1, 31, 31, 128)
after Conv_3     (1, 31, 31, 128)
after Conv_4     (1, 31, 31, 256)
after max_pool_1 (1, 30, 30, 256)
after Conv_5     (1, 30, 30, 512)
after max_pool_2 (1, 29, 29, 512)
after Conv_6     (1, 29, 29, 512)
after Conv_7     (1, 29, 29, 512)
after max_pool_3 (1, 26, 26, 512)
```

**Reply by samuela** [reply]
AFAICT this is consistent with the OG cifar10-fast pytorch version however. It pools [like this](https://github.com/davidcpage/cifar10-fast/blob/f79e8b89bb2994b5caaaf7289c68903d15a37c2e/dawn_utils.py#L29) and the default [MaxPool2d in PyTorch](https://pytorch.org/docs/stable/generated/torch.nn.MaxPool2d.html) seems to be padding=0, which would be consistent with [linen's defaults](https://github.com/google/flax/blob/6ab8cb27b5792fdfbd4b17e6dc1f5074f9ba721f/flax/linen/pooling.py#L91-L104). 

However, I am noticing that the PyTorch version uses a max pool kernel size of 4, whereas the JAX implementation uses 2. I could've sworn it was 2, but perhaps not. Running the JAX version with kernel size 4 takes 18s/epoch, an improvement but still nowhere near <3s/epoch. My JAX code is available [here](https://gist.github.com/samuela/6d704930b9cc5a88f149d692983f2c22).

EDIT: found why I thought it [was MaxPool2d(2)](https://github.com/samuela/cifar10-fast/blob/a660ec033899d10820a1e57b480f949d6033716e/dawn_utils.py#L21). The pytorch code is a bit messy; it's not entirely clear what is happening.

**Reply by samuela** [reply]
Just for kicks I tried changing the max_pool's to have padding="SAME" with the hypothesis that cuDNN/XLA was picking crappy kernels on oddly shaped images, but sadly that slowed training down to 31s/epoch.

**Reply by samuela** [reply]
Using
```py
x = nn.max_pool(x, (4, 4), strides=(2, 2), padding="SAME")
```
brings it down to 2.2s/epoch. But that's totally different from what the pytorch version is doing AFAIU. That corresponds to shapes:
```
[nix-shell:~/dev/research/lottery]$ nixGLNvidia-510.47.03 python cifar10_dawnbench.py
init             (1, 32, 32, 3)
after Conv_0     (1, 32, 32, 64)
after Conv_1     (1, 32, 32, 128)
after max_pool_0 (1, 16, 16, 128)
after Conv_2     (1, 16, 16, 128)
after Conv_3     (1, 16, 16, 128)
after Conv_4     (1, 16, 16, 256)
after max_pool_1 (1, 8, 8, 256)
after Conv_5     (1, 8, 8, 512)
after max_pool_2 (1, 4, 4, 512)
after Conv_6     (1, 4, 4, 512)
after Conv_7     (1, 4, 4, 512)
after max_pool_3 (1, 2, 2, 512)
```

**Reply by samuela** [reply]
~The more I read this weird pytorch code, the less confident I am that I've transcribed it correctly into jax/flax. It's deeply confusing and self-contradictory in some places...~

EDIT: Nope I was right all along.

**Reply by SandSnip3r** [reply]
Lol sorry to send you down this rabbit hole. I feel your pain with the pytorch code. It's very bizarre to read. I started looking into this and was doubting that it matched up, but I'm not super confident since I'm not a practicing ml engineer who's familiar with these things.

**Reply by samuela** [reply]
Yeah, I think this code is especially hairy... I went down the rabbit hole and I'm happy to report that AFAICT my original code was correct all along, with the minor exception that the pytorch version doesn't use a bias term on the final layer. So that's surely costing the JAX version a whole 1e-6ms/epoch.

For anyone interested, there are a few confusing things in reading the pytorch code:
* It uses a custom NN library implemented in `torch_backend.py`
* Said NN library makes use of the unfortunate fact that dictionaries are ordered in Python.
* Max pooling with kernel width 2 is used throughout the intermediate blocks but max pooling with kernel width 4 is used just before flattening at the end of the network.

[Here](https://github.com/samuela/cifar10-fast/blob/a660ec033899d10820a1e57b480f949d6033716e/dawn_utils.py#L21-L38
) is the entry point for those brave souls who would like to stare into the abyss and experience the abyss staring back at you.

**Reply by SandSnip3r** [reply]
The Pytorch version is using `MaxPool2d(2)`. This creates a 2x2 max pool which strides 2-at-a-time with no padding. This will result in the dimensions halving after each conv&pool. I added a little print in torch_backend.py's `Network::forward` and here is some size information after each step:
```
Node:Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
  torch.Size([32, 64, 32, 32])
Node:Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
  torch.Size([32, 128, 32, 32])
Node:MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  torch.Size([32, 128, 16, 16])
Node:Identity()
  torch.Size([32, 128, 16, 16])
Node:Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
  torch.Size([32, 128, 16, 16])
Node:Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
  torch.Size([32, 128, 16, 16])
Node:Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
  torch.Size([32, 256, 16, 16])
Node:MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  torch.Size([32, 256, 8, 8])
Node:Conv2d(256, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
  torch.Size([32, 512, 8, 8])
Node:MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  torch.Size([32, 512, 4, 4])
Node:Identity()
  torch.Size([32, 512, 4, 4])
Node:Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
  torch.Size([32, 512, 4, 4])
Node:Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
  torch.Size([32, 512, 4, 4])
Node:MaxPool2d(kernel_size=4, stride=4, padding=0, dilation=1, ceil_mode=False)
  torch.Size([32, 512, 1, 1])
Node:Flatten()
  torch.Size([32, 512])
Node:Linear(in_features=512, out_features=10, bias=False)
  torch.Size([32, 10])
Node:Mul()
  torch.Size([32, 10])
Node:CrossEntropyLoss()
  torch.Size([32])
Node:Correct()
  torch.Size([32])
  ```

Also, it's worth noticing that the JAX version is using NHWC data format, while PyTorch is using NCHW. This can be a source of other performance differences.

**Reply by samuela** [reply]
Ahhh great catch! That's very sneaky. I had no idea that pytorch by default sets stride = kernel size... Sure as shit, that brings down the time to 2.02s/epoch! Thank you so much @SandSnip3r! Very tricky indeed...

**Reply by SandSnip3r** [reply]
I updated your code to set stride to `(2, 2)` and it does speed it up. However, I don't see your version converging on a good test accuracy at the same rate. Do you see the same?

**Reply by samuela** [reply]
That's entirely possible. My version doesn't do any batch norm, data aug, etc at the moment. I'm confident that with all the same tweaks it would get similar results. If not, then there are certainly other correctness bugs lurking, but hopefully no more performance ones!
